src/Microsoft.PowerShell.Commands.Utility/commands/utility/TestJsonCommand.cs

```diff
@@ -32,33 +36,98 @@ public class TestJsonCommand : PSCmdlet
+        {
+            if (e is TargetInvocationException)
+            {
+                ExceptionDispatchInfo.Capture(e.InnerException).Throw();
+            }
+            else
+            {
+                ExceptionDispatchInfo.Capture(e).Throw();
+            }
+            return true;
+        }
+
+        /// <summary>
+        /// Prepare a JSON schema.
         /// </summary>
         protected override void BeginProcessing()
         {
-            if (Schema != null)
+            try
             {
-                try
+                if (Schema != null)
                 {
-                    _jschema = JsonSchema.FromJsonAsync(Schema).Result;
+                    try
+                    {
+                        _jschema = JsonSchema.FromJsonAsync(Schema).Result;
+                    }
+                    // Even if only one exception is thrown, it is still wrapped in an AggregateException exception
+                    // https://docs.microsoft.com/en-us/dotnet/standard/parallel-programming/exception-handling-task-parallel-library
+                    catch (AggregateException ae)
+                    {
+                        ae.Handle(UnwrapException);
+                    }
                 }
-                catch (Exception exc)
+                else if (SchemaFile != null)
                 {
-                    Exception exception = new Exception(TestJsonCmdletStrings.InvalidJsonSchema, exc);
-                    ThrowTerminatingError(new ErrorRecord(exception, "InvalidJsonSchema", ErrorCategory.InvalidData, null));
+                    try
+                    {
+                        string resolvedpath = Context.SessionState.Path.GetUnresolvedProviderPathFromPSPath(SchemaFile);
+                        _jschema = JsonSchema.FromFileAsync(resolvedpath).Result;
+                    }
+                    catch (AggregateException ae)
+                    {
+                        ae.Handle(UnwrapException);
+                    }
                 }
             }
+            // Handle exceptions related to file access to provide more specific error message
+            // https://docs.microsoft.com/en-us/dotnet/standard/io/handling-io-errors
+            catch (Exception e) when (
+                e is IOException ||
+                e is UnauthorizedAccessException ||
+                e is NotSupportedException ||
+                e is SecurityException
+            )
+            {
+                // Do we really need to wrap exception? Not doing this provides more clear error message upfront.
+                // E.g.: "'{}'|Test-Json -SchemaFile c:" results in "Test-Json : Access to the path 'C:\' is denied".
+                Exception exception = new Exception("JSON schema file open failure", e); // TODO: Add resource string
+                ThrowTerminatingError(new ErrorRecord(exception, "JsonSchemaFileOpenFailure", ErrorCategory.OpenError, null));
```

Comment:
Please write non-terminating error.

Reply:
I'm not sure how non-terminating error (`WriteError` ?) is supposed to work. If we don't terminate here, then we need to somehow handle this in the `ProcessRecord`, and it doesn't make sense to run `ProcessRecord` at all since the scheme file can't be accessed.

P.S> There is infinitesimal that file will become accessible on the subsequent pipeline iterations, but...

Reply:
Do WriteError end return false.

Reply:
This cmdlet's verb is `Test` - the primary output mode should be true / false, not throwing errors. We're not asserting that the json is valid and erroring otherwise, we're testing it. 😕

Reply:
If we have a problem with schema file we should write a terminating error.
For JSON test we should return true/false but in interactive session I'd expect to see non-terminating error by default. It is less typing in interactive session to see useful result. In script mode user can suppress errors with -ErrorActoion SilentlyContinue - more typing but it is not too important as for interactive session.

Reply:
@iSazonov Ok, I see how it can be done, thanks.

@vexx32 Returning true/false is fine when we're talking about the resource itself (JSON), but if schema is invalid (or not found) shouldn't we treat it as an invalid input and throw?

E.g.: `Test-Connection` will throw if it can't resolve host, `Test-ModuleManifest` throws if path is invalid or not pointing to the `psd1` file..

Reply:
Oh, this is if the schema is totally invalid / can't be found? 

Then yeah, ignore me, this is totally appropriate. Sorry! 😊

Reply:
I'd like to continue working on this. 

> @iSazonov
> If we have a problem with schema file we should write a terminating error.

This is exactly what the code here is doing - does it means that it's ok?